dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/queryengine/TaskCreator.scala

```diff
@@ -0,0 +1,125 @@
+      val param = result.path.head.node.asInstanceOf[MethodParameterIn]
+      if (result.callSiteStack.isEmpty) {
+        // Case 1
+        paramToArgs(param).map { arg =>
+          ReachableByTask(arg, sources, new ResultTable, result.path, result.callDepth + 1)
+        }
+      } else {
+        // Case 2
+        val newCallSiteStack = result.callSiteStack.clone()
+        val callSite         = newCallSiteStack.pop()
+        paramToArgs(param).filter(x => x.inCall.exists(c => c == callSite)).map { arg =>
+          ReachableByTask(arg, sources, new ResultTable, result.path, result.callDepth + 1, newCallSiteStack)
+        }
+      }
+    }
+  }
+
+  /** Returns only those results that start at a parameter node.
+    */
+  private def startsAtParameter(results: Vector[ReachableByResult]) = {
+    results.collect { case r: ReachableByResult if r.path.head.node.isInstanceOf[MethodParameterIn] => r }
+  }
+
+  /** For a given parameter of a method, determine all corresponding arguments at all call sites to the method.
+    */
+  private def paramToArgs(param: MethodParameterIn): List[Expression] =
+    NoResolve
+      .getMethodCallsites(param.method)
+      .to(Traversal)
+      .collectAll[Call]
+      .argument(param.index)
+      .l
+
+  /** Create new tasks from all results that end in an output argument, including return arguments. In this case, we
+    * want to traverse to corresponding method output parameters and method return nodes respectively.
+    */
+  private def tasksForUnresolvedOutArgs(results: Vector[ReachableByResult]): Vector[ReachableByTask] = {
+
+    val outArgsAndCalls = results
+      .map(x => (x, x.outputArgument, x.path, x.callDepth))
+      .distinct
+
+    val forCalls = outArgsAndCalls.flatMap { case (result, outArg, path, callDepth) =>
+      val outCall = outArg.collect { case n: Call => n }
+
+      val methodReturns = outCall.toList
+        .flatMap(x => NoResolve.getCalledMethods(x).methodReturn.map(y => (x, y)))
+        .to(Traversal)
+
+      methodReturns.flatMap { case (call, methodReturn) =>
+        val returnStatements = methodReturn._reachingDefIn.toList.collect { case r: Return => r }
+        if (returnStatements.isEmpty) {
+          val newPath       = path
+          val callSiteStack = result.callSiteStack.clone()
+          callSiteStack.push(call)
+          List(ReachableByTask(methodReturn, sources, new ResultTable, newPath, callDepth + 1, callSiteStack))
+        } else {
+          returnStatements.map { returnStatement =>
+            val newPath       = Vector(PathElement(methodReturn)) ++ path
+            val callSiteStack = result.callSiteStack.clone()
+            callSiteStack.push(call)
+            ReachableByTask(returnStatement, sources, new ResultTable, newPath, callDepth + 1, callSiteStack)
+          }
+        }
+      }
+    }
+
+    val forArgs = outArgsAndCalls.flatMap { case (result, args, path, callDepth) =>
+      args.toList.flatMap { arg =>
```

Comment:
If you want to get rid of unchecked `.asInstanceOf` casts
```suggestion
      args.toList.flatMap { case (arg: Expression) =>
```